File: cdm/cpp/patient/actions/SEPericardialEffusion.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

#include "stdafx.h"
#include "patient/actions/SEPericardialEffusion.h"
#include "properties/SEScalarVolumePerTime.h"
#include "io/protobuf/PBPatientActions.h"

SEPericardialEffusion::SEPericardialEffusion() : SEPatientAction()
{
  m_EffusionRate = nullptr;
}

SEPericardialEffusion::~SEPericardialEffusion()
{
  Clear();
}

void SEPericardialEffusion::Clear()
{
  SEPatientAction::Clear();
  SAFE_DELETE(m_EffusionRate);
}

void SEPericardialEffusion::Copy(const SEPericardialEffusion& src)
{
  PBPatientAction::Copy(src, *this);
}

bool SEPericardialEffusion::IsValid() const
{
  return SEPatientAction::IsValid() && HasEffusionRate();
}

bool SEPericardialEffusion::IsActive() const
{
  return IsValid() ? !m_EffusionRate->IsZero() : false;
}

bool SEPericardialEffusion::HasEffusionRate() const
{
  return m_EffusionRate == nullptr ? false : m_EffusionRate->IsValid();
}
SEScalarVolumePerTime& SEPericardialEffusion::GetEffusionRate()
{
  if (m_EffusionRate == nullptr)
    m_EffusionRate = new SEScalarVolumePerTime();
  return *m_EffusionRate;
}
double SEPericardialEffusion::GetEffusionRate(const VolumePerTimeUnit& unit) const
{
  if (m_EffusionRate == nullptr)
    return SEScalar::dNaN();
  return m_EffusionRate->GetValue(unit);
}

void SEPericardialEffusion::ToString(std::ostream &str) const
{
  str << "Patient Action : Pericardial Effusion"; 
  if(HasComment())
    str<<"\n\tComment: "<<m_Comment;
  str << "\n\tEffusion Rate: "; HasEffusionRate() ? str << *m_EffusionRate : str << "NaN";
  str << std::flush;
}

File: cdm/csharp/patient/actions/SEPericardialEffusion.cs
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

namespace Pulse.CDM
{
  public class SEPericardialEffusion : SEPatientAction
  {
    protected SEScalarVolumePerTime effusion_rate;

    public SEPericardialEffusion()
    {
      effusion_rate = null;
    }

    public override void Clear()
    {
      base.Clear();
      if (effusion_rate != null)
        effusion_rate.Invalidate();
    }

    public override bool IsValid()
    {
      return HasEffusionRate();
    }

    public bool HasEffusionRate()
    {
      return effusion_rate == null ? false : effusion_rate.IsValid();
    }
    public SEScalarVolumePerTime GetEffusionRate()
    {
      if (effusion_rate == null)
        effusion_rate = new SEScalarVolumePerTime();
      return effusion_rate;
    }
  }
}
